src/main/java/slimeknights/tconstruct/plugin/jei/casting/MaterialCastingRecipeMaker.java

```diff
@@ -0,0 +1,62 @@
+import java.util.Objects;
+
+@NoArgsConstructor(access = AccessLevel.PRIVATE)
+public final class MaterialCastingRecipeMaker {
+  public static List<ItemCastingRecipe.Basin> createMaterialCastingBasinRecipes() {
+    List<ItemCastingRecipe.Basin> recipes = new ArrayList<>();
+    List<MaterialCastingRecipe.Basin> materialCastingRecipes = RecipeUtil.getRecipes(Minecraft.getInstance().world.getRecipeManager(), RecipeTypes.CASTING_BASIN, MaterialCastingRecipe.Basin.class);
+    String group = "jei.material.casting.";
+    for (IMaterial material : MaterialRegistry.getInstance().getMaterials()) {
+      if (material.getFluid() == Fluids.EMPTY) continue;
+      for (MaterialCastingRecipe.Basin basin : materialCastingRecipes) {
+        IMaterialItem result = basin.getResult();
+        ItemStack partStack = result.getItemstackWithMaterial(material);
+        String path = String.format("jei.material.casting.%s.%s", Objects.requireNonNull(result.asItem().getRegistryName()).getPath(), Objects.requireNonNull(material.getFluid().getRegistryName()).getPath());
+        ResourceLocation id = new ResourceLocation(TConstruct.modID, path);
+        ItemCastingRecipe.Basin tableRecipe = new ItemCastingRecipe.Basin(id, group + Objects.requireNonNull(result.asItem().getRegistryName()).getPath(), basin.getCast(), FluidIngredient.of(material.getFluid(), basin.getFluidAmount()), partStack, ICastingRecipe.calcCoolingTime(material.getTemperature(), basin.getFluidAmount()), basin.isConsumed(), basin.switchSlots());
+        recipes.add(tableRecipe);
+      }
+    }
+    return recipes;
+  }
+
+  public static List<ItemCastingRecipe.Table> createMaterialCastingTableRecipes() {
+    List<ItemCastingRecipe.Table> recipes = new ArrayList<>();
+    List<MaterialCastingRecipe.Table> materialCastingRecipes = RecipeUtil.getRecipes(Minecraft.getInstance().world.getRecipeManager(), RecipeTypes.CASTING_TABLE, MaterialCastingRecipe.Table.class);
+    String group = "jei.material.casting.";
+    for (IMaterial material : MaterialRegistry.getInstance().getMaterials()) {
+      if (material.getFluid() == Fluids.EMPTY) continue;
+      for (MaterialCastingRecipe.Table table : materialCastingRecipes) {
+        IMaterialItem result = table.getResult();
+        ItemStack partStack = result.getItemstackWithMaterial(material);
+        String path = String.format("jei.material.casting.%s.%s", Objects.requireNonNull(result.asItem().getRegistryName()).getPath(), Objects.requireNonNull(material.getFluid().getRegistryName()).getPath());
+        ResourceLocation id = new ResourceLocation(TConstruct.modID, path);
+        ItemCastingRecipe.Table tableRecipe = new ItemCastingRecipe.Table(id, group + Objects.requireNonNull(result.asItem().getRegistryName()).getPath(), table.getCast(), FluidIngredient.of(material.getFluid(), table.getFluidAmount()), partStack, ICastingRecipe.calcCoolingTime(material.getTemperature(), table.getFluidAmount()), table.isConsumed(), table.switchSlots());
```

Comment:
Leave the group alone, those are supposed to be the same for similar recipes.

Reply:
so group as `""` then? Mostly was just adapting JEIs setup.